crates/llm-ls/src/document.rs

```diff
@@ -148,73 +158,422 @@ impl Document {
+                    PositionEncodingKind::UTF8 => change_start_line_cu_idx,
+                    PositionEncodingKind::UTF16 => {
+                        change_start_line.char_to_utf16_cu(change_start_line_char_idx)
+                    }
+                    PositionEncodingKind::UTF32 => change_start_line_char_idx,
+                };
+                let change_end_line_byte_idx = match same_line && same_character {
+                    true => change_start_line_byte_idx,
+                    false => match position_encoding {
+                        PositionEncodingKind::UTF8 => change_end_line_cu_idx,
+                        PositionEncodingKind::UTF16 => {
+                            change_end_line.char_to_utf16_cu(change_end_line_char_idx)
+                        }
+                        PositionEncodingKind::UTF32 => change_end_line_char_idx,
+                    },
+                };
+
+                self.text
+                    .remove(change_start_doc_char_idx..change_end_doc_char_idx);
+                self.text.insert(change_start_doc_char_idx, &change.text);
+
+                if let Some(tree) = &mut self.tree {
+                    // 6. Compute the byte index into the new end line where the
+                    // change ends. Required for tree-sitter.
+                    let change_new_end_line_idx = self
+                        .text
+                        .byte_to_line(change_start_doc_byte_idx + change.text.len());
+                    let change_new_end_line_byte_idx =
+                        change_start_doc_byte_idx + change.text.len();
+
+                    // 7. Construct the tree-sitter edit. We stay mindful that
+                    // tree-sitter Point::column is a byte offset.
+                    let edit = InputEdit {
+                        start_byte: change_start_doc_byte_idx,
+                        old_end_byte: change_end_doc_byte_idx,
+                        new_end_byte: change_start_doc_byte_idx + change.text.len(),
+                        start_position: Point {
+                            row: change_start_line_idx,
+                            column: change_start_line_byte_idx,
+                        },
+                        old_end_position: Point {
+                            row: change_end_line_idx,
+                            column: change_end_line_byte_idx,
+                        },
+                        new_end_position: Point {
+                            row: change_new_end_line_idx,
+                            column: change_new_end_line_byte_idx,
+                        },
+                    };
+
+                    tree.edit(&edit);
+
+                    self.tree = Some(self
+                                .parser
+                                .parse(self.text.to_string(), Some(tree))
+                                .expect("parse should always return a tree when the language was set and no timeout was specified"));
```

Comment:
```suggestion
                                .expect("parse should always return a tree when the language was set and no timeout was specified"));
```
This should be handled properly

Reply:
I did not understand what you meant by this, how do you want me to handle this case ?

Reply:
I don't think this should cause any issues. The tree is `None` if no language was set and we don't set a timeout. I believe these are the only two ways the `parse` method can panic. In any case @jeremyelalouf, you can always propagate the error upwards.

Reply:
I do agree that the use of any `expect`s, `unwrap`s, or `panic!` should be minimized, as mentioned. 

> It's probably better to propagate an error message to the client in case we add a timeout in the future.

As you suggested, Diego, if a potential timeout in the future was set, having a predefined custom error type, such as `Error::TreeSitterParseError`, is advantageous.

This would provide additional details for the user.